### PR TITLE
Articles can be assigned categories 

### PR DESCRIPTION
**Pull request by jyuw**
### Duplication explanation 
Due to a database issue in the code base in PR #17 and many many hours spent trying to rectify it, we decided it would be faster to recreate the code base from scratch since it was the weekend and we wanted to continue working instead of being stuck for another entire day. Pulling the code from #17 resulted in that entire branch's database becoming unusable as well, so it wasn't worth salvaging. We therefore made a team decision to scrap it and start again.

#### Description
Changes proposed in this pull request:

* Articles must be given a category
* Categories will be shown on the index page

#### What I have learned working on this feature:
* how to configure databasecleaner
* making join tables

Screenshots:
screen shot 2018-03-17 at 8 31 43 pm
screen shot 2018-03-17 at 8 31 49 pm

### CODE REVIEW

**Comment by tochman** [comment]
Please dive into markdown and reformat your PR description

File: features/support/env.rb
Comment (tochman) [inline features/support/env.rb]: IF you need this (and I'm not sure you do, `cucumber/rails' might be requiring this already) you need to move it out of Coveralls ' way.

File: features/support/env.rb
Comment (tochman) [inline features/support/env.rb]: Right, If you require db cleaner on line 2, why do you require it here again?

File: features/support/env.rb
Comment (tochman) [inline features/support/env.rb]: I don't mind this change, but I would like a clarification on why you made it, please.

File: features/user_should_be_able_to_create_articles.feature
Comment (tochman) [inline features/user_should_be_able_to_create_articles.feature]: Please try to use capitalized labels in your scenarios

File: features/user_should_be_able_to_edit_articles.feature
Comment (tochman) [inline features/user_should_be_able_to_edit_articles.feature]: Use line breaks between steps

File: app/controllers/articles_controller.rb
Comment (tochman) [inline app/controllers/articles_controller.rb]: What does this do for you?

File: app/controllers/articles_controller.rb
Comment (tochman) [inline app/controllers/articles_controller.rb]: I would assume you need to save the instance once you added the categories, right?

File: app/views/articles/edit.html.erb
Comment (tochman) [inline app/views/articles/edit.html.erb]: Please clarify why you use `u` as variable name?

File: app/views/articles/new.html.erb
Comment (tochman) [inline app/views/articles/new.html.erb]: Please note that there is a lot of code repetition on this form. Perhaps consider extracting all fields into a partial?

File: features/articles_belong_to_categories.feature
Comment (tochman) [inline features/articles_belong_to_categories.feature]: ` And I am on the "Index" page` seems like you are telling Capybara to visit a page, not asserting that you are on a page. Is it an assertion or is it a command?

File: features/support/env.rb
Comment (jyuw) [reply features/support/env.rb]: The database wasn't being cleaned properly without these changes, we kept getting this 
screen shot 2018-03-18 at 9 05 06 am
But we'll remove one of the `require` lines

File: app/models/article.rb
Comment (tochman) [inline app/models/article.rb]: There's a lot of code repetition on LOC's 2, 3 and 4. Please use the shorter
```
validates_presence_of :title, :body, :categories
```

File: spec/factories/articles.rb
Comment (tochman) [inline spec/factories/articles.rb]: This is not the way you work with associations in FactoryBot. Please take a closer look at https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations

File: features/articles_belong_to_categories.feature
Comment (holgertidemand) [reply features/articles_belong_to_categories.feature]: We want it to visit the page! :) 
We wanted to see if the category was displayed on the index page.
It's within scope & we thought that a new Scenario was overkill, so we just added two lines to this Scenario. 
But I can see that we might want to add a Scenario?! 

what do you think?

File: features/support/env.rb
Comment (tochman) [reply features/support/env.rb]: Okay. I'm pretty convinced toy can remove them both.

**Review by tochman — CHANGES_REQUESTED** [review]
Please address my comments.

**Comment by tochman** [comment]
We would love to see more git collaboration on this teams PR's If you work together, as you do on this PR, please take turns on writing code and making commits.

File: spec/factories/articles.rb
Comment (holgertidemand) [reply spec/factories/articles.rb]: This was the main reason the CI build failed! We took it out and it worked! Thanks for the link!

Comment (jyuw) [reply spec/factories/articles.rb]: We tried several ways of making the association to category, including the association line from the link, but it always says this:
screen shot 2018-03-18 at 9 22 37 am

Using line 5 was the only way we were able to get the tests to pass locally, but it breaks on semaphore which we thought was the problem.

File: features/articles_belong_to_categories.feature
Comment (tochman) [reply features/articles_belong_to_categories.feature]: I see. Thx for the clarification. A bit tricky. At the very least I would change `And` to `When` and perhaps rephrase to `When I visit the "Index" page` to be more clear. 

Also, hold back on the exclamation marks in your comments @holgertidemand They can be easily misunderstood.

File: app/controllers/articles_controller.rb
Comment (holgertidemand) [reply app/controllers/articles_controller.rb]: The params returns a array with two indices. The first one is always empty. 
``` params[:article][:categories] = ["", "id_category"] ``` 
We tried a couple of ways to remove it.

File: app/controllers/articles_controller.rb
Comment (holgertidemand) [reply app/controllers/articles_controller.rb]: We save it on line 8

File: app/views/articles/edit.html.erb
Comment (holgertidemand) [reply app/views/articles/edit.html.erb]: we will change it to 'c' for category!

File: features/support/env.rb
Comment (jyuw) [reply features/support/env.rb]: Then we get this
screen shot 2018-03-18 at 9 36 08 am

File: features/articles_belong_to_categories.feature
Comment (holgertidemand) [reply features/articles_belong_to_categories.feature]: We will refactor this! And no more '!' :)

File: features/support/env.rb
Comment (jyuw) [reply features/support/env.rb]: It works with only the ` DatabaseCleaner.clean_with :truncation` line so I'll get rid of the rest, but we need the truncation one

File: spec/factories/articles.rb
Comment (tochman) [reply spec/factories/articles.rb]: I would try something like this to start with:
```ruby
FactoryBot.define do
  factory :article do
    title "MyString"
    body "MyText"
    categories {[FactoryBot.create(:category)]}
  end
end
```

File: app/controllers/articles_controller.rb
Comment (tochman) [reply app/controllers/articles_controller.rb]: I don’t understand

File: app/views/articles/edit.html.erb
Comment (tochman) [inline app/views/articles/edit.html.erb]: We don't set id's with capitals

Comment (tochman) [reply app/views/articles/edit.html.erb]: If the test is not passing, use `downcase` in your step definition

File: app/controllers/articles_controller.rb
Comment (tochman) [inline app/controllers/articles_controller.rb]: Since you provide the ability to to add multiple categories to the article, you probably want to rename this method to reflect that. `add_categories_to_article`

File: app/controllers/articles_controller.rb
Comment (tochman) [inline app/controllers/articles_controller.rb:24]: I still don't get what it does for you so I can't approve it.

File: app/controllers/articles_controller.rb
Comment (tochman) [reply app/controllers/articles_controller.rb]: Okay, thx.

File: app/views/articles/edit.html.erb
Comment (tochman) [inline app/views/articles/edit.html.erb]: Again, the form fields contain a lot of repetition. Is a partial a solution? `form_with` will know what action you perform, so there's no need to add `value:` to the submit form. And if you have `@article` defined on the `:new` and `:edit` actions, you should be able to use a partial.

Comment (tochman) [reply app/views/articles/edit.html.erb]: Also, I'm not a great fan of single letter variable names.

Comment (tochman) [reply app/views/articles/edit.html.erb]: I would also set an `@categories = Category.all` instance variable in your controller to avoid calling the db from your view. (Not a requirement though)

File: app/views/articles/new.html.erb
Comment (tochman) [inline app/views/articles/new.html.erb]: See comments above.

File: app/views/layouts/_nav.html.erb
Comment (tochman) [inline app/views/layouts/_nav.html.erb:2]: Is this a placeholder for now?

File: app/views/layouts/application.html.erb
Comment (tochman) [inline app/views/layouts/application.html.erb]: Formatting seems off

File: spec/factories/articles.rb
Comment (tochman) [inline spec/factories/articles.rb:5]: 👍🏼

**Review by tochman — CHANGES_REQUESTED** [review]
Please see comments.

File: features/support/env.rb
Comment (tochman) [reply features/support/env.rb]: Yes, you do.

File: app/controllers/articles_controller.rb
Comment (holgertidemand) [reply app/controllers/articles_controller.rb]: If we loop ``` params[:article][:categories]  ``` as is, we will have a empty string in our ```article.categories```. 
When we try to id that empty string to display the categories it thorws an error. 
We remove that empty string with the command ``` params[:article][:categories].shift ```
Do you have a suggestion on another solution for this issue?

File: app/views/articles/edit.html.erb
Comment (holgertidemand) [reply app/views/articles/edit.html.erb]: okay

File: app/controllers/articles_controller.rb
Comment (holgertidemand) [reply app/controllers/articles_controller.rb:24]: Read above answer

File: app/views/articles/edit.html.erb
Comment (holgertidemand) [reply app/views/articles/edit.html.erb]: IMO this is easier to read this then to create a partial for just two line of code!

Comment (holgertidemand) [reply app/views/articles/edit.html.erb]: Instance variable makes since.

File: app/views/layouts/_nav.html.erb
Comment (holgertidemand) [reply app/views/layouts/_nav.html.erb:2]: Yes. We dont have a show for categories yet.

Comment (holgertidemand) [reply app/views/layouts/_nav.html.erb:2]: outside of scope

**Comment by tochman** [comment]
Please update the PR description with the background of WHY this was created as a duplicate of https://github.com/CraftAcademy/dream_team_publications/pull/17 I think a clarification is in place. Once that is done, this one is ready to merge IMO. I'll lift my request for changes at that point. 

The team need to have a discussion about the workflow and usage of Git. There was NO need to create a new PR from what I can see in the original one.

File: app/controllers/articles_controller.rb
Comment (tochman) [reply app/controllers/articles_controller.rb:24]: I did.  That answer is NOT bringing clarity to the issue imo. I still don’t get it. Will move on though.

**Review by tochman — APPROVED** [review]

**Review by larathordar — APPROVED** [review]
Great job!

**Pull request merged**
